swim/disseminator.go

```diff
@@ -109,63 +109,88 @@ func (d *disseminator) FullSync() (changes []Change) {
 	return changes
 }
 
-func (d *disseminator) IssueAsSender() []Change {
-	return d.issueChanges(nil)
+// IssueAsSender collects all changes a node needs when sending a ping or
+// ping-req. The second return value is a callback that raises the piggyback
+// counters of the given changes.
+func (d *disseminator) IssueAsSender() (changes []Change, bumpPiggybackCounters func([]Change)) {
+	return d.issueChanges(), d.bumpPiggybackCounters
 }
 
-func (d *disseminator) IssueAsReceiver(senderAddress string,
-	senderIncarnation int64, senderChecksum uint32) ([]Change, bool) {
+func (d *disseminator) bumpPiggybackCounters(changes []Change) {
+	d.Lock()
+	for _, change := range changes {
+		c, ok := d.changes[change.Address]
+		if !ok {
+			continue
+		}
 
-	filter := func(c *pChange) bool {
-		return senderIncarnation == c.SourceIncarnation &&
-			senderAddress == c.Source
+		c.p += 1
+		if c.p >= d.maxP {
+			delete(d.changes, c.Address)
+		}
 	}
+	d.Unlock()
+}
 
-	changes := d.issueChanges(filter)
+// IssueAsReceiver collects all changes a node needs when responding to a ping
+// or ping-req. Unlike IssueAsSender, IssueAsReceiver automatically increments
+// the piggyback counters. The second return value indicates whether a full sync
+// is triggered.
+func (d *disseminator) IssueAsReceiver(
+	senderAddress string,
+	senderIncarnation int64,
+	senderChecksum uint32) (changes []Change, fullSync bool) {
 
-	if len(changes) > 0 {
-		return changes, false
-	} else if d.node.memberlist.Checksum() != senderChecksum {
-		d.node.emit(FullSyncEvent{senderAddress, senderChecksum})
+	changes = d.issueChanges()
 
-		d.node.log.WithFields(log.Fields{
-			"localChecksum":  d.node.memberlist.Checksum(),
-			"remote":         senderAddress,
-			"remoteChecksum": senderChecksum,
-		}).Info("full sync")
+	// filter out changes that came from the sender previously
+	changes = d.filterChangesFromSender(changes, senderAddress, senderIncarnation)
+
+	d.bumpPiggybackCounters(changes)
 
-		return d.FullSync(), true
+	if len(changes) > 0 || d.node.memberlist.Checksum() == senderChecksum {
+		return changes, false
 	}
 
-	return []Change{}, false
+	d.node.emit(FullSyncEvent{senderAddress, senderChecksum})
+
+	d.node.log.WithFields(log.Fields{
+		"localChecksum":  d.node.memberlist.Checksum(),
+		"remote":         senderAddress,
+		"remoteChecksum": senderChecksum,
+	}).Info("full sync")
+
+	return d.FullSync(), true
 }
 
-func (d *disseminator) issueChanges(filter func(*pChange) bool) (changes []Change) {
+// filterChangesFromSender returns changes that didn't originate at the sender.
+// Attention, this function reorders the underlaying input array.
+func (d *disseminator) filterChangesFromSender(cs []Change, source string, incarnation int64) []Change {
+	for i := 0; i < len(cs); i++ {
+		if incarnation == cs[i].SourceIncarnation && source == cs[i].Source {
+			d.node.emit(ChangeFilteredEvent{cs[i]})
+			cs[i], cs[len(cs)-1] = cs[len(cs)-1], cs[i]
```

Comment:
Gotcha. We don't have need for it now, but I prefer to not overwrite changes in the underlying array, which you still have access to. I figured that reordering is not a concern since the order is non-deterministic to begin with. This saves us some memory allocation.

If you're unaware of how go does slices internally http://blog.golang.org/go-slices-usage-and-internals.

It explains that if you do:
b := a[:10]
where a has length 20. Slice a still exists and still has length 20. b just has access to the first 10 elements of a. Mutating elements of b will therefor mutate a. 

There are some subtleties in where you next do:
b = append(b, 999)
b now needs to grow since it's a its max capacity. The go runtime allocates a bigger array somewhere and copies the contents to it. If we now mutate elements of b, a is no longer mutated.